README.md

[![Arduino CI](https://github.com/RobTillaart/TSL260R/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/TSL260R/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/TSL260R/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/TSL260R/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/TSL260R/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/TSL260R/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/TSL260R.svg?maxAge=3600)](https://github.com/RobTillaart/TSL260R/releases)


# TSL260R

Arduino library for the TSL260R IR to voltage convertor


## Description

The TSL260R (TSL261R, TSL262R) is a IR sensor that outputs a voltage depending on the irradiation.

This library does convert the output voltage to uW/cm2.

As the type sensor differ by sensitivity the library has three distinct classes.
The table below is an approximation for the max irradiation at 3.3 Volt (output).
For an Arduino UNO 3.3 V is about 650 ADC steps. 
When using e.g. an external 16 bit ADS1115, one definitely has far more steps.

|  Type     |  max uW/cm2  |
|:---------:|:------------:|
|  TSL260R  |     30       |
|  TSL261R  |     70       |
|  TSL262R  |    125       |


**Warning** this library is experimental so you should use it with care.
It is written on the datasheet, and I have no hardware yet to test it. 
Of course I am very interested in your experiences and feedback to improve
the library.


## Hardware Connection

#### Power supply

The maximum output voltage depends on the power supply voltage.
This implies that the output range (uW/cm2) depends on power supply voltage.
To maximize the measurement range a voltage of at leat 4.5 V is advised.

See datasheet figure 14: Maximum Output Voltage vs Supply Voltage


#### Schema

Always check datasheet 

```
//  Front view 
//
//  PIN 1 - GND
//  PIN 2 - VDD      2.7 V .. 5.5 V
//  PIN 3 - SIGNAL   voltage out
```


## Interface

#### Internal ADC

- **TSL260R(uint8_t pin, uint16_t maxADC, float voltage)** Constructor when using an 
internal ADC and just one sample to measure the output voltage of the sensor.
  - pin = analogRead() pin
  - maxADC = max value of the internal ADC, UNO = 1023.
  - voltage = max voltage of the internal ADC.
- **TSL261R(uint8_t pin, uint16_t maxADC, float voltage)** idem for TSL261R.
- **TSL262R(uint8_t pin, uint16_t maxADC, float voltage)** idem for TSL262R.
- **float irradiance()** returns the irradiance in uW/cm2.
Uses the analogRead() of the internal ADC. 
**Fails** by returning 0 when object is created with the other constructor.


#### External ADC

- **TSL260R()** constructor when using an external ADC or more than one internal samples
to measure the voltage.
When using this constructor one cannot use the parameterless **irradiance()**, see above.
- **TSL261R()** idem for TSL261R.
- **TSL262R()** idem for TSL262R.
- **float irradiance(float voltage)** returns the irradiance in uW/cm2 based upon voltage 
parameter. Does not use an internal analogRead(). 
Note the user can average 8 internal ADC measurements to reduce noise and then use 
this function based upon the average voltage.


#### WaveLength

- **void setWaveLength(uint16_t waveLength = 940)** sets the wave length so the conversion 
can use a correction factor. 
At 900 - 940 nm the wave length correction factor == 1.0.
Wave length should be between 830 and 1100.
- **uint16_t getWaveLength()** returns the configured wave length.
- **float getWaveLengthFactor()** returns the wave length correction factor. 
Note the sensor is most sensitive around 940 nm. See datasheet.
- **calculateWaveLengthFactor(uint16_t waveLength)** calculates the factor to compensate 
for less sensitivity at other wave lengths.
E.g. if the sensor is 0.5 x as sensitive at a given wave length the factor should be 2. 


#### Calibration

Since version 0.1.2 the following functions are added to calibrate the irradiance formula
to some extend. The formula is ```irradiance = AA * voltage + BB```.

See datasheet figure 12: Output Voltage vs Irradiance

Use with care.

- **void     setAA(float aa)** set a new value for AA.
- **float    getAA()** return the current value.
- **void     setBB(float bb)** set a new value for BB.
- **float    getBB()** return the current value.


## Operations

See examples.


## Future

#### must
- improve documentation
- buy hardware (where)
- test test test test


#### should
- extend unit tests
- write examples
- fix the dependency of **irradiance()**
  - derived class?
- optimize code
- 

#### could
- test with different IR LEDS (e.g. remote)
- Normalized Output Voltage vs Angular Displacement correction
  - figure 11
- temperature correction
- 



CHANGELOG.md
# Change Log TSL260R

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.1.3] - 2023-02-18

- update readme.md
- update GitHub actions
- update license 2023
- minor edits


## [0.1.2] - 2022-11-27

- add getter / setter for \_aa and \_bb
- add unit tests for get/set above
- update comments
- update readme.md
- update examples
- fix space in .cpp name

## [0.1.1] - 2022-11-27

- update documentation
- add analogRead Constructor
- add irradiance() for internal analogRead()
- fix camelCase WaveLength
- update unit tests.
- add extra example.
- add keywords.txt

## [0.1.0] - 2022-11-25

- initial version - written on datasheet only
  Need to buy hardware for testing.



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2022-11-25
// PURPOSE: unit tests for the temperature library
//          https://github.com/RobTillaart/TSL260R
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)


#include <ArduinoUnitTests.h>


#include "Arduino.h"
#include "TSL260R.h"


unittest_setup()
{
  fprintf(stderr, "TSL260R_LIB_VERSION: %s\n", (char *) TSL260R_LIB_VERSION);
}


unittest_teardown()
{
}


unittest(test_constructor_I)
{
  TSL260R TSL0;
  TSL261R TSL1;
  TSL262R TSL2;

  assertEqual(940, TSL0.getWavelength() );
  assertEqual(940, TSL1.getWavelength() );
  assertEqual(940, TSL2.getWavelength() );

  assertEqualFloat(1.0, TSL0.getWaveLengthFactor(), 0.001);
  assertEqualFloat(1.0, TSL1.getWaveLengthFactor(), 0.001);
  assertEqualFloat(1.0, TSL2.getWaveLengthFactor(), 0.001);
}


unittest(test_constructor_II)
{
  TSL260R TSL0(14, 1023, 5.0);

  assertEqual(940, TSL0.getWavelength() );
  assertEqualFloat(1.0, TSL0.getWaveLengthFactor(), 0.001);
}


unittest(test_irradiance)
{
  TSL260R TSL0;
  TSL261R TSL1;
  TSL262R TSL2;

  //  see output example 1
  assertEqualFloat(6.985,  TSL0.irradiance(0.700), 0.001);
  assertEqualFloat(16.305, TSL1.irradiance(0.700), 0.001);
  assertEqualFloat(77.000, TSL2.irradiance(0.700), 0.001);
}


unittest(test_calculateWaveLengthFactor)
{
  TSL260R TSL;

  //  check the internal table 
  assertEqualFloat(100,   TSL.calculateWaveLengthFactor(800), 0.1);
  assertEqualFloat(10,    TSL.calculateWaveLengthFactor(830), 0.01);
  assertEqualFloat(1.25,  TSL.calculateWaveLengthFactor(880), 0.001);
  assertEqualFloat(1.0,   TSL.calculateWaveLengthFactor(900), 0.001);
  assertEqualFloat(1.0,   TSL.calculateWaveLengthFactor(940), 0.001);
  assertEqualFloat(2.0,   TSL.calculateWaveLengthFactor(990), 0.001);
  assertEqualFloat(3.333, TSL.calculateWaveLengthFactor(1050), 0.001);
  assertEqualFloat(10,    TSL.calculateWaveLengthFactor(1100), 0.01);
  assertEqualFloat(100,   TSL.calculateWaveLengthFactor(1150), 0.1);
}



unittest_main()


// --------


TSL260R.h
#pragma once
//
//    FILE: TSL260R.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.3
//    DATE: 2022-11-25
// PURPOSE: library for the TSL260R IR to voltage convertor
//     URL: https://github.com/RobTillaart/TSL260R


#define TSL260R_LIB_VERSION             (F("0.1.3"))

#include "Arduino.h"


class TSL260R
{
public:

  //  for internal ADC
  TSL260R(uint8_t pin, uint16_t maxADC, float voltage);
  //  for external ADC
  TSL260R();

  //  use internal ADC, defined in constructor
  //  uses a single analogRead() => defined by constructor.
  float    irradiance();
  //  use external (or internal) ADC voltage.
  //  e.g to average multiple analogRead() calls
  float    irradiance(float voltage);

  //  range: 800 - 1150 ==> 940 = default.
  void     setWaveLength(uint16_t waveLength = 940);
  uint16_t getWaveLength();
  float    getWaveLengthFactor();

  //  useful for debugging too
  float    calculateWaveLengthFactor(uint16_t waveLength);

  //  irradiance parameters
  //  only change these with care.
  void     setAA(float aa);
  float    getAA();
  void     setBB(float aa);
  float    getBB();


protected:
  uint8_t  _pin              = 0;
  float    _voltagePerStep   = 0;
  uint16_t _waveLength       = 940;
  float    _waveLengthFactor = 1;
  //       _aa and _bb are defined in constructor;
  //       need getter / setter to adjust values runtime
  float    _aa;
  float    _bb;
  float    multiMap(float value, float * _in, float * _out, uint8_t size);
};


///////////////////////////////////////////////////////
//
//  DERIVED CLASSES
//
class TSL261R : public TSL260R
{
public:
  TSL261R();
  TSL261R(uint8_t pin, uint16_t maxADC, float voltage);
};


class TSL262R : public TSL260R
{
public:
  TSL262R();
  TSL262R(uint8_t pin, uint16_t maxADC, float voltage);
};


// -- END OF FILE --




TSL260R.cpp
//
//    FILE: TSL260R.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.3
//    DATE: 2022-11-25
// PURPOSE: library for the TSL260R IR to voltage convertor
//     URL: https://github.com/RobTillaart/TSL260R


#include "TSL260R.h"


TSL260R::TSL260R(uint8_t pin, uint16_t maxADC, float voltage)
{
  _pin = pin;
  _voltagePerStep = voltage / maxADC;
  //  datasheet page 9 figure 12
  //  irradiance parameters
  _aa = 10.0067;
  _bb = -0.02013423;
  //  waveLength parameters
  _waveLength       = 940;
  _waveLengthFactor = 1.0;
}


TSL260R::TSL260R()
{
  TSL260R(0, 1, 0); //  prevent divide by zero
}


float TSL260R::irradiance(float voltage)
{
  float value = _aa * voltage + _bb;
  if (_waveLengthFactor != 1.0) value *= _waveLengthFactor;
  return value;
}


float TSL260R::irradiance()
{
  if (_voltagePerStep <= 0) return 0;
  float voltage = analogRead(_pin) * _voltagePerStep;
  return irradiance(voltage);
}


void TSL260R::setWaveLength(uint16_t waveLength)
{
  _waveLength = waveLength;
  _waveLengthFactor = calculateWaveLengthFactor(_waveLength);
}


uint16_t TSL260R::getWaveLength()
{
  return _waveLength;
}


float TSL260R::getWaveLengthFactor()
{
  return _waveLengthFactor;
}


float TSL260R::calculateWaveLengthFactor(uint16_t waveLength)
{
  //  figure 12 datasheet
  //  940 nm is reference 1.000
  //  remaining is linear interpolated between points in the graph
  float in[]  = { 800,  830, 880, 900, 940, 990, 1050, 1100, 1150};
  float out[] = { 0.01, 0.1, 0.8, 1.0, 1.0, 0.5, 0.3,  0.1,  0.01};
  //  the correction factor is reciprocate as sensor is less sensitive.
  return 1.0 / multiMap(waveLength, in, out, 9);
}


///////////////////////////////////////////////////////
//
//  irradiance parameters
//
void TSL260R::setAA(float aa) { _aa = aa; }
float TSL260R::getAA()        { return _aa; }

void TSL260R::setBB(float bb) { _bb = bb; }
float TSL260R::getBB()        { return _bb; }


///////////////////////////////////////////////////////
//
//  PRIVATE
//
float TSL260R::multiMap(float value, float * _in, float * _out, uint8_t size)
{
  //  take care the value is within range
  //  value = constrain(value, _in[0], _in[size-1]);
  if (value <= _in[0]) return _out[0];
  if (value >= _in[size-1]) return _out[size-1];

  //  search right interval
  uint8_t pos = 1;  // _in[0] already tested
  while(value > _in[pos]) pos++;

  //  this will handle all exact "points" in the _in array
  if (value == _in[pos]) return _out[pos];

  //  interpolate in the right segment for the rest
  uint8_t pos1 = pos - 1;
  return (value - _in[pos1]) * (_out[pos] - _out[pos1]) / (_in[pos] - _in[pos1]) + _out[pos1];
}



///////////////////////////////////////////////////////
//
//  TSL261R
//
TSL261R::TSL261R() : TSL260R()
{
  //  datasheet page 9
  //  irradiance parameters
  _aa = 23.34564;
  _bb = -0.03692;
}


TSL261R::TSL261R(uint8_t pin, uint16_t maxADC, float voltage) : TSL260R(pin, maxADC, voltage)
{
  //  datasheet page 9
  //  irradiance parameters
  _aa = 23.34564;
  _bb = -0.03692;
}


///////////////////////////////////////////////////////
//
//  TSL262R
//
TSL262R::TSL262R() : TSL260R()
{
  //  datasheet page 9
  //  irradiance parameters
  _aa = 110;
  _bb = 0;
}


TSL262R::TSL262R(uint8_t pin, uint16_t maxADC, float voltage) : TSL260R(pin, maxADC, voltage)
{
  //  datasheet page 9
  //  irradiance parameters
  _aa = 110;
  _bb = 0;
}


//  -- END OF FILE --



TSL260R .cpp
//
//    FILE: TSL260R.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.1
//    DATE: 2022-11-25
// PURPOSE: library for the TSL260R IR to voltage convertor


#include "TSL260R.h"


TSL260R::TSL260R(uint8_t pin, uint16_t maxADC, float voltage)
{
  _pin = pin;
  _voltagePerStep = voltage / maxADC;
  //  datasheet page 9 figure 12
  //  voltage parameters
  _aa = 10.0067;
  _bb = -0.02013423; 
  //  waveLength parameters
  _waveLength       = 940; 
  _waveLengthFactor = 1.0;
}


TSL260R::TSL260R()
{
  TSL260R(0, 1, 0); //  prevent divide by zero
  //  datasheet page 9
  //  voltage parameters
  _aa = 10.0067;
  _bb = -0.02013423; 
  //  waveLength parameters
  _waveLength       = 940; 
  _waveLengthFactor = 1.0;
}


float TSL260R::irradiance(float voltage)
{
  float value = _aa * voltage + _bb;
  if (_waveLengthFactor != 1.0) value *= _waveLengthFactor;
  return value;
}


float TSL260R::irradiance()
{
  float voltage = analogRead(_pin) * _voltagePerStep;
  return irradiance(voltage);
}


void TSL260R::setWaveLength(uint16_t waveLength)
{
  _waveLength = waveLength;
  _waveLengthFactor = calculateWaveLengthFactor(_waveLength);
}


uint16_t TSL260R::getWaveLength()
{
  return _waveLength; 
}


float TSL260R::getWaveLengthFactor() 
{
  return _waveLengthFactor; 
}


float TSL260R::calculateWaveLengthFactor(uint16_t waveLength)
{
  //  figure 12 datasheet
  //  940 nm is reference 1.000
  //  remaining is linear interpolated between points in the graph
  float in[]  = { 800,  830, 880, 900, 940, 990, 1050, 1100, 1150};
  float out[] = { 0.01, 0.1, 0.8, 1.0, 1.0, 0.5, 0.3,  0.1,  0.01};
  //  the correction factor is reciprocate as sensor is less sensitive.
  return 1.0 / multiMap(waveLength, in, out, 9);
}


float TSL260R::multiMap(float value, float * _in, float * _out, uint8_t size)
{
  //  take care the value is within range
  //  value = constrain(value, _in[0], _in[size-1]);
  if (value <= _in[0]) return _out[0];
  if (value >= _in[size-1]) return _out[size-1];

  //  search right interval
  uint8_t pos = 1;  // _in[0] already tested
  while(value > _in[pos]) pos++;

  //  this will handle all exact "points" in the _in array
  if (value == _in[pos]) return _out[pos];

  //  interpolate in the right segment for the rest
  uint8_t pos1 = pos - 1;
  return (value - _in[pos1]) * (_out[pos] - _out[pos1]) / (_in[pos] - _in[pos1]) + _out[pos1];
}



///////////////////////////////////////////////////////
//
//  TSL261R
//
TSL261R::TSL261R() : TSL260R()
{
  //  datasheet page 9
  //  voltage parameters
  _aa = 23.34564;
  _bb = -0.03692;
  //  waveLength parameters
  _waveLength       = 940; 
  _waveLengthFactor = 1.0;
}


TSL261R::TSL261R(uint8_t pin, uint16_t maxADC, float voltage) : TSL260R(pin, maxADC, voltage)
{
  //  datasheet page 9
  //  voltage parameters
  _aa = 23.34564;
  _bb = -0.03692;
  //  waveLength parameters
  _waveLength       = 940; 
  _waveLengthFactor = 1.0;
}


///////////////////////////////////////////////////////
//
//  TSL262R
//
TSL262R::TSL262R() : TSL260R()
{
  //  datasheet page 9
  //  voltage parameters
  _aa = 110;
  _bb = 0;
  //  waveLength parameters
  _waveLength       = 940; 
  _waveLengthFactor = 1.0;
}



TSL262R::TSL262R(uint8_t pin, uint16_t maxADC, float voltage) : TSL260R(pin, maxADC, voltage)
{
  //  datasheet page 9
  //  voltage parameters
  _aa = 110;
  _bb = 0;
  //  waveLength parameters
  _waveLength       = 940; 
  _waveLengthFactor = 1.0;
}



//  -- END OF FILE --

